### ISSUE TITLE
False negative in WhitespaceAfter/WhitespaceAround with tab after reserved word

### ISSUE THREAD

**Issue by VeniceWish** [checkstyle/checkstyle#17719]
I have read check documentation: 
https://checkstyle.sourceforge.io/checks/whitespace/whitespaceafter.html
https://checkstyle.sourceforge.io/checks/whitespace/whitespacearound.html
https://checkstyle.sourceforge.io/checks/whitespace/filetabcharacter.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

From: https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace
> Separating any reserved word, such as if, for or catch, from an open parenthesis (() that follows it on that line.

Code: 
```java
class Test {
  void testTab() {
    // There is a Tab between "for" and "("
    for	(int i = 0; i < 5; i++) { // Expected 2 violations but get 1
      System.out.println(i);
    }
  }
}

```

Cli: 
```cmd
$ java -jar checkstyle-10.26.1-all.jar -c google_checks.xml Test.java 
Starting audit...
[WARN] Test.java:4:8: Line contains a tab character. [FileTabCharacter]
Audit done.
```

Formatter:
```cmd
$ java -jar google-java-format-1.28.0-all-deps.jar Test.java > FormattedCode.java
$ diff -Naru Test.java FormattedCode.java
--- Test.java   2025-09-01 21:55:47.629476200 +0800
+++ FormattedCode.java  2025-09-01 22:17:27.049535100 +0800
@@ -1,7 +1,7 @@
 class Test {
   void testTab() {
     // There is a Tab between "for" and "("
-    for        (int i = 0; i < 5; i++) { // Expected 2 Violation but get 1
+    for (int i = 0; i < 5; i++) { // Expected 2 Violation but get 1
       System.out.println(i);
     }
   }
```

---

When a reserved word `for` is followed by a tab before the `(`, Checkstyle only reports `FileTabCharacter`.
But it does not report `WhitespaceAfter` / `WhitespaceAround`, even though the style guide requires a single space.
So we only get 1 violation instead of 2 (tab + missing space).

---

**Comment by romani** [comment]
All Checks are independent from each other, so they do validation completely separately.
We combined Checks in Google style config to get coverage.
We consider that if at least one violation happened, it fully covering style guide.
So in current example, there is single tab, and user get violation on it. What Check is doing it, does not matter much if it works well.

So I can not justify that others Checks need to do more violations.
I can not approve this issue.

Please provide more reasons, or we can close issue.

**Comment by VeniceWish** [comment]
Thank you for the response!

I raised this issue mainly from the perspective of **automated fixing workflows**. If all violations could be reported in a single run, then tools (or users) can fix them directly based on the output without needing to rerun Checkstyle.

Currently, only the `FileTabCharacter` check reports a violation when a tab is used instead of a space. Based on that message alone, an automated fixer would likely just delete the tab, but would not know to insert a space, since `WhitespaceAfter` or `WhitespaceAround` are silent in that scenario.

Of course, running Checkstyle again after the first fix will catch the missing space. So this is solvable, but it requires at least two iterations.

What you said does make sense—especially in an iterative style of fixing where it's enough to know there's some violation on the line.

Thanks again for the clarification!

**Comment by romani** [comment]
It depends on what auto fixer you use. But  yes, it is not rare case when double run is required. It mostly a concerns of auto formatter.

What do you use?
We started https://github.com/checkstyle/checkstyle-openrewrite-recipes initiative.
